WIP: Replace Karma with Jest

nathanstilwell opened the pull request:
**This is currently a work in progress.**

This change seeks to replace the test runner [Karma](https://karma-runner.github.io/latest/index.html) with [Jest](https://jestjs.io/). Jest has become our standard test runner across the Cockroach Labs org, so we are seeking to unify the testing tools for UI development. Jest comes with many tools built in and so will replace many additional tools we are using with Karma, such as [Mocha](https://mochajs.org/), [Chai](https://www.chaijs.com/), [`assert`](https://github.com/browserify/commonjs-assert), and [Sinon](https://sinonjs.org/).

Jest uses [`describe`](https://jestjs.io/docs/api#describename-fn) the same as Mocha, and has support for `it` (although prefers [test](https://jestjs.io/docs/api#testname-fn-timeout) for individual tests) so the structure of tests shouldn't change. Assertions are replaces with the [`Expect`](https://jestjs.io/docs/expect) global and a set of [matchers](https://jestjs.io/docs/using-matchers) so tests have been refactored in this change. Jest also has its own [mocking function](https://jestjs.io/docs/mock-function-api) so Sinon spies will be replaces as well.

sjbarag committed:
Add jest and friends to db-console

had to do this by manually editing package.json to ensure we got the
same versions as in cluster-ui

sjbarag committed:
allow files that match /.*\.test.*/ in pkg/ui/

it's idiomatic for JS files

sjbarag committed:
copy jest configs from cluster-ui as a starting point

it'll probably break

sjbarag committed:
snapshot: results of jest-codemod 'src/**/*.spec.(ts|tsx)' for jasmine (global usage)

sjbarag committed:
enable esModuleInterop in tsconfig.json

For some reason, db-console used to work without it, but jest testing
requires it.  TODO: figure out why

Presented as "_ is undefined" after `import _ from "lodash";`, which
makes sense.

sjbarag committed:
snapshot
<pkg>: <short description - lowercase, no final period>

<what was there before: Previously, ...>
<why it needed to change: This was inadequate because ...>
<what you did about it: To address this, this patch ...>

Release justification:
Release note (<category, see below>): <what> <show> <why>

Commit:
ui: linting jest config

Release justification:
Release note: None

Commit:
ui: splitting src paths in tsconfig

Getting the ts-jest warning, `Mapping only to first target of "src/*" because it has more than one (2)`,
which seems to be causing some import issues.

Release justification:
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/80728)
<!-- Reviewable:end -->

Commit:
ui: another run of jest-codemods

This time on raw chai assertions, which seemed to get most of them.

Release justification:
Release note: None

Commit:
ui: Change `jsx` in db-console tsconfig to `react`

This seems to remove `unexpected token` errors for jsx tags in Jest /
ts-jest.

Release justification:
Release note: None

nathanstilwell commented:
Closed this PR in place of https://github.com/cockroachdb/cockroach/pull/82161
